### ISSUE TITLE
More informative errors for Prompter

### ISSUE THREAD

**Issue by RyanMarten** [bespokelabsai/curator#41]
Code:
```
from bespokelabs import curator

poet = curator.Prompter(
    prompt_func=lambda: "Write a poem about the beauty of computer science",
    parse_func=lambda _, response: f"Poem:\n{response}",
    model_name="gpt-4o-mini",
)

poem = poet()
print(poem["response"][0])
```

Error:
```
Traceback (most recent call last):
  File "/Users/ryan/curator/examples/poem_test.py", line 9, in <module>
    poem = poet()
           ^^^^^^
  File "/Users/ryan/curator/src/bespokelabs/curator/prompter/prompter.py", line 78, in __call__
    return self._completions(self._request_processor, dataset)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryan/curator/src/bespokelabs/curator/prompter/prompter.py", line 159, in _completions
    dataset = request_processor.run(
              ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryan/curator/src/bespokelabs/curator/request_processor/openai_online_request_processor.py", line 197, in run
    dataset = self.create_dataset_files(dataset, working_dir, prompt_formatter)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryan/curator/src/bespokelabs/curator/request_processor/base_request_processor.py", line 276, in create_dataset_files
    writer.finalize()
  File "/Users/ryan/curator/.venv/lib/python3.12/site-packages/datasets/arrow_writer.py", line 594, in finalize
    self.write_examples_on_file()
  File "/Users/ryan/curator/.venv/lib/python3.12/site-packages/datasets/arrow_writer.py", line 440, in write_examples_on_file
    if all(isinstance(row[0][col], (pa.Array, pa.ChunkedArray)) for row in self.current_examples):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryan/curator/.venv/lib/python3.12/site-packages/datasets/arrow_writer.py", line 440, in <genexpr>
    if all(isinstance(row[0][col], (pa.Array, pa.ChunkedArray)) for row in self.current_examples):
                      ~~~~~~^^^^^
TypeError: string indices must be integers, not 'str'
```

**Comment by RyanMarten** [comment]
Another one, where the cache gets in a broken state:
```
Traceback (most recent call last):
  File "/Users/ryan/curator/examples/poem_test.py", line 9, in <module>
    poem = poet()
           ^^^^^^
  File "/Users/ryan/curator/src/bespokelabs/curator/prompter/prompter.py", line 78, in __call__
    return self._completions(self._request_processor, dataset)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryan/curator/src/bespokelabs/curator/prompter/prompter.py", line 159, in _completions
    dataset = request_processor.run(
              ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryan/curator/src/bespokelabs/curator/request_processor/openai_online_request_processor.py", line 197, in run
    dataset = self.create_dataset_files(dataset, working_dir, prompt_formatter)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryan/curator/src/bespokelabs/curator/request_processor/base_request_processor.py", line 229, in create_dataset_files
    return Dataset.from_file(dataset_file)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryan/curator/.venv/lib/python3.12/site-packages/datasets/arrow_dataset.py", line 740, in from_file
    table = ArrowReader.read_table(filename, in_memory=in_memory)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryan/curator/.venv/lib/python3.12/site-packages/datasets/arrow_reader.py", line 329, in read_table
    return table_cls.from_file(filename)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryan/curator/.venv/lib/python3.12/site-packages/datasets/table.py", line 1017, in from_file
    table = _memory_mapped_arrow_table_from_file(filename)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryan/curator/.venv/lib/python3.12/site-packages/datasets/table.py", line 63, in _memory_mapped_arrow_table_from_file
    opened_stream = _memory_mapped_record_batch_reader_from_file(filename)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryan/curator/.venv/lib/python3.12/site-packages/datasets/table.py", line 49, in _memory_mapped_record_batch_reader_from_file
    return pa.ipc.open_stream(memory_mapped_stream)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryan/curator/.venv/lib/python3.12/site-packages/pyarrow/ipc.py", line 190, in open_stream
    return RecordBatchStreamReader(source, options=options,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryan/curator/.venv/lib/python3.12/site-packages/pyarrow/ipc.py", line 52, in __init__
    self._open(source, options=options, memory_pool=memory_pool)
  File "pyarrow/ipc.pxi", line 1006, in pyarrow.lib._RecordBatchStreamReader._open
  File "pyarrow/error.pxi", line 155, in pyarrow.lib.pyarrow_internal_check_status
  File "pyarrow/error.pxi", line 92, in pyarrow.lib.check_status
pyarrow.lib.ArrowInvalid: Tried reading schema message, was null or length 0
```

**Comment by RyanMarten** [comment]
How we deal with None as parse_func and None as Dataset is currently unintuitive since the interface behaves differently.

**Comment by RyanMarten** [comment]
Fixed with #61

**Comment by RyanMarten** [comment]
Fixed with #61
